### PR TITLE
Binary sensor map implementation

### PR DESCRIPTION
**Pull request by mvturnho**
## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mvturnho** [comment]
@OttoWinter can we also include this component? I have implemented this in several applications so it is well tested.

File: esphome/components/binary_sensor_map/sensor.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+
+from esphome.components import sensor, binary_sensor
+from esphome.const import CONF_ID, CONF_CHANNELS, CONF_CHANNEL, CONF_VALUE, CONF_TYPE
+
+DEPENDENCIES = ['binary_sensor']
+
+binary_sensor_map_ns = cg.esphome_ns.namespace('binary_sensor_map')
+BinarySensorMap = binary_sensor_map_ns.class_('BinarySensorMap', cg.Component)
```
Comment (OttoWinter) [inline esphome/components/binary_sensor_map/sensor.py]: This also inherits from `sensor.Sensor`

File: esphome/components/binary_sensor_map/sensor.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+SensorMapType = binary_sensor_map_ns.enum('SensorMapType')
+SENSOR_MAP_TYPES = {
+    'GROUP': SensorMapType.BINARY_SENSOR_MAP_TYPE_GROUP,
+}
+
+entry = {
+    cv.Required(CONF_CHANNEL): cv.use_id(binary_sensor.BinarySensor),
+    cv.Required(CONF_VALUE): cv.All(cv.positive_int, cv.Range(min=0, max=255)),
```
Comment (OttoWinter) [inline esphome/components/binary_sensor_map/sensor.py]: Why is this restricted to integers? The struct accepts a float.

File: esphome/components/binary_sensor_map/binary_sensor_map.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+   * @param value  The value this binary_sensor represents
+   */
+  void add_channel(binary_sensor::BinarySensor *sensor, float value);
+  void set_sensor_type(BinarySensorMapType sensor_type);
+
+ protected:
+  std::vector<BinarySensorMapChannel *> sensors_{};
+  BinarySensorMapType sensor_type_{BINARY_SENSOR_MAP_TYPE_GROUP};
+  bool last_any_active_{false};
+  // this gives max 46 channels per binary_sensor_map
```
Comment (OttoWinter) [inline esphome/components/binary_sensor_map/binary_sensor_map.h]: 64, not 46 😉

**Review by OttoWinter — APPROVED** [review]
Yeah sure, this can get into 1.13

I have updated the PR with some things that I think could have been slightly improved (see commit message for more info).

In this state, I think it is done. I've also tested the changes with an ESP32 touch sensor and can confirm it works.

**Pull request merged**